lib/semantics/resolve-labels.cc

```diff
@@ -379,118 +381,62 @@ class ParseTreeAnalyzer {
-    if (!firstNameNoneOrBothEqual(
-            std::get<parser::Statement<parser::EndModuleStmt>>(module.t)
-                .statement.v,
-            std::get<parser::Statement<parser::ModuleStmt>>(module.t)
-                .statement.v)) {
-      errorHandler_.Say(currentPosition_, "END MODULE name mismatch"_err_en_US)
-          .Attach(
-              std::get<parser::Statement<parser::ModuleStmt>>(module.t).source,
-              "mismatched MODULE"_en_US);
-    }
+    CheckOptionalName<parser::ModuleStmt>("MODULE", module,
+        std::get<parser::Statement<parser::EndModuleStmt>>(module.t));
   }
+
   // C1569
   void Post(const parser::SeparateModuleSubprogram &separateModuleSubprogram) {
-    if (!firstNameNoneOrBothEqual(
-            std::get<parser::Statement<parser::EndMpSubprogramStmt>>(
-                separateModuleSubprogram.t)
-                .statement.v,
-            std::get<parser::Statement<parser::MpSubprogramStmt>>(
-                separateModuleSubprogram.t)
-                .statement.v)) {
-      errorHandler_
-          .Say(currentPosition_, "END MODULE PROCEDURE name mismatch"_err_en_US)
-          .Attach(std::get<parser::Statement<parser::MpSubprogramStmt>>(
-                      separateModuleSubprogram.t)
-                      .source,
-              "mismatched MODULE PROCEDURE"_en_US);
-    }
+    CheckOptionalName<parser::MpSubprogramStmt>("MODULE PROCEDURE",
+        separateModuleSubprogram,
+        std::get<parser::Statement<parser::EndMpSubprogramStmt>>(
+            separateModuleSubprogram.t));
   }
+
   // C1401
   void Post(const parser::MainProgram &mainProgram) {
-    if (std::get<std::optional<parser::Statement<parser::ProgramStmt>>>(
-            mainProgram.t)
-            .has_value()) {
-      auto &programStmt{
-          std::get<std::optional<parser::Statement<parser::ProgramStmt>>>(
-              mainProgram.t)
-              ->statement};
-      auto &endProgramStmt{
-          std::get<parser::Statement<parser::EndProgramStmt>>(mainProgram.t)
-              .statement};
-      if (endProgramStmt.v.has_value()) {
-        if (programStmt.v.ToString() != endProgramStmt.v->ToString()) {
-          errorHandler_
-              .Say(currentPosition_, "END PROGRAM name mismatch"_err_en_US)
-              .Attach(
-                  std::get<
-                      std::optional<parser::Statement<parser::ProgramStmt>>>(
-                      mainProgram.t)
-                      ->source,
-                  "mismatched PROGRAM"_en_US);
-        }
-      }
-    } else {
-      if (std::get<parser::Statement<parser::EndProgramStmt>>(mainProgram.t)
-              .statement.v.has_value()) {
-        errorHandler_.Say(currentPosition_,
-            parser::MessageFormattedText{
-                "END PROGRAM cannot have a program-name"_err_en_US});
+    auto &endProgramStmt{
+        std::get<parser::Statement<parser::EndProgramStmt>>(mainProgram.t)
+            .statement};
```

Comment:
It seems like this could be simplified further by use GetStmtName() here too. Maybe:
```
const parser::CharBlock *endName{GetStmtName(
    std::get<parser::Statement<parser::EndProgramStmt>>(mainProgram.t))};
if (const auto &program{
    std::get<std::optional<parser::Statement<parser::ProgramStmt>>>(mainProgram.t)}) {
  const parser::CharBlock *name{GetStmtName(*program)};
...
```

Reply:
Good idea, and I think that it can be even further simplified.